src/Params/WaveTable.h

```diff
@@ -0,0 +1,659 @@
+
+    // ringbuffer
+    AbstractRingbuffer::size_t size_freqs() const { return freqs.size(); }
+    AbstractRingbuffer::size_t size_semantics() const { return data.size(); }
+    AbstractRingbuffer::size_t write_space_semantics(tensor_size_t freq_idx) const { return data.ringbuffers[freq_idx].write_space(); }
+    AbstractRingbuffer::size_t write_space_delayed_semantics(tensor_size_t freq_idx) const { return data.ringbuffers[freq_idx].write_space_delayed(); }
+    AbstractRingbuffer::size_t write_pos_semantics(tensor_size_t freq_idx) const { return data.ringbuffers[freq_idx].write_pos(); }
+    AbstractRingbuffer::size_t write_pos_delayed_semantics(tensor_size_t freq_idx) const { return data.ringbuffers[freq_idx].write_pos_delayed(); }
+    void inc_read_pos_semantics(tensor_size_t freq_idx) { data.ringbuffers[freq_idx].inc_read_pos(); }
+    void inc_write_pos_semantics(tensor_size_t freq_idx, AbstractRingbuffer::size_t amnt) { data.ringbuffers[freq_idx].inc_write_pos(amnt); }
+    void inc_write_pos_delayed_semantics(tensor_size_t freq_idx, AbstractRingbuffer::size_t amnt) { data.ringbuffers[freq_idx].inc_write_pos_delayed(amnt); }
+    void dump_rb(tensor_size_t freq_idx) const {
+#ifdef DBG_WAVETABLES
+        data.ringbuffers[freq_idx].AbstractRingbuffer::dump();
+#else
+        (void)freq_idx;
+#endif
+    }
+
+    // timestamp/outdated stuff
+    // only valid for the *current* table, not for the next one
+    void set_outdated(int timestamp) { assert(timestamp != timestamp_current); timestamp_requested = timestamp; }
+    bool outdated() const { return timestamp_requested != timestamp_current; }
+    void set_current_timestamp(int timestamp) { timestamp_current = timestamp; }
+    bool is_correct_timestamp(int timestamp) const { return timestamp_requested == timestamp; }
+    int debug_get_timestamp_requested() const { return timestamp_requested; }
+    int debug_get_timestamp_current() const { return timestamp_current; }
+
+    // update request stuff
+    void require_update_request() { m_require_update_request = true; }
+    bool update_request_required() const { return m_require_update_request; }
+    void update_request_sent() { m_require_update_request = false; }
+
+    // change stamp stuff
+    void setChangeStamp(unsigned to) { change_stamp = to; }
+    unsigned changeStamp() const { return change_stamp; }
+
+    // getting data
+    float get_freq(tensor_size_t freq_idx) const { return freqs[freq_idx]; }
+    IntOrFloat get_sem(tensor_size_t sem_idx) const { return semantics[sem_idx]; }
+    const Tensor1<IntOrFloat>* const* get_semantics_addr() const { return &semantics_addr; }
+    const Tensor1<float32>* const* get_freqs_addr() const { return &freqs_addr; }
+    //! Find best index in freq tensor to play frequency freq
+    tensor_size_t findBestIndex(float freq) const;
+    //! Return sample slice for given frequency
+    const Tensor1<float32> &getNextWaveToConsume(float32 freq);
```

Comment:
ok, this is the expected float32 arrays I was expecting over the ringbuffer.

Reply:
It returns them. The implementation of this function indeed does something like

```c++
return data[rb.read_pos()][bestI]
```

Which shows that the ringbuffer just contains the abstract indices, but no float buffers. Instead, we use the ringbuffer indices to receive the float buffers from the `Tensor3ForWaveTable data`.

Reply:
Does this imply repeated calls into getNextWaveToConsume(float32 freq) with a fixed freq will attempt to get different waveforms?

If so then the ringbuffer is a datastructure which is providing a mechanism to fetch least recently used (and hopefully not used at all) notes?

Reply:
> Does this imply repeated calls into getNextWaveToConsume(float32 freq) with a fixed freq will attempt to get different waveforms?

Mostly yes, except of border cases. It is not called in WT modulation. If it is called with random in use, then the answer is yes. However, if no random is in effect (the default for ad synth, random is only active if OscilGen's phase or amplitude randomness are enabled), then the size of this ringbuffer is 1 and it intentionally never gets emptied - as can be seen in `getNextWaveToConsume`.

> If so then the ringbuffer is a datastructure which is providing a mechanism to fetch least recently used (and hopefully not used at all) notes?

Assuming the ringbuffer size is >1. It fetches least recently produced notes, but usually, they were never recently used. One time only consumption and then they are removed from the ringbuffer. MW will refill the buffer with new random seeds.

Reply:
ok, then there might be some current issues in the case where oscilgen is a random generator and unison is turned on. That would result in fetching a ton of different waves where unison typically is one base wave played sampled by a bunch of slightly different resamplers.

Reply:
Like already mentioned in another comment, for me it looks like the current git master does not play one wave at different frequencies, but rather calls `OscilGen::get` for each of the unison frequencies, each time with different random - which is why the WT code keeps this behavior. At least, this is how I understand https://github.com/zynaddsubfx/zynaddsubfx/blob/806674849b276a560d2996f25ba0ddbfe50838ef/src/Synth/ADnote.cpp#L455C1-L459C35

Reply:
oh goodness, that feels like the current git master is doing something that's occasionally insanely expensive on note on. I thought we were currently at one get per oscillator which could be bad, but unison_size Oscilgen::get calls feels like that's just asking for xruns.

In that case feel free to disregard this point since your solution replicates prior behavior, but it's way cheaper :)

Reply:
Issue on master will be tracked on #61 , so resolving this comment.